bin/varnishd/cache/cache_vrt.c

```diff
@@ -339,6 +339,119 @@ VRT_CollectString(VRT_CTX, const char *p, ...)
 	return (b);
 }
 
+/*--------------------------------------------------------------------
+ * Collapse STRANDS into the space provided, or return NULL
+ */
+
+char *
+VRT_Strands(char *d, unsigned dl, VCL_STRANDS s)
+{
+	char *b, *e;
+	unsigned x;
+
+	b = d;
+	e = b + dl;
+	for (int i = 0; i < s->n && b < e; i++)
+		if (s->p[i] != NULL && *s->p[i] != '\0') {
+			x = strlen(s->p[i]);
+			if (b + x < e)
+				memcpy(b, s->p[i], x);
+			b += x;
+		}
+	if (b >= e)
+		return (NULL);
+	*b++ = '\0';
+	return (b);
+}
+
+/*--------------------------------------------------------------------
+ * Copy and merge STRANDS into a workspace.
+ */
+
+VCL_STRING
+VRT_StrandsWS(struct ws *ws, const char *h, VCL_STRANDS s)
+{
+	char *b, *e;
+	const char *q = NULL;
+	VCL_STRING r;
+	unsigned u, x;
+	int i;
+
+	u = WS_Reserve(ws, 0);
+
+	for (i = 0; i < s->n; i++)
+		if (s->p[i] != NULL && *s->p[i] != '\0') {
+			q = s->p[i];
+			break;
+		}
+
+	if (h != NULL && q == NULL && WS_Inside(ws, h, NULL)) {
+		WS_Release(ws, 0);
+		return (h);
+	}
+
+	if (h == NULL && q != NULL && s->n == 1 && WS_Inside(ws, q, NULL)) {
```

Comment:
what `if s->n == 2 && s->p[0] == NULL`
I think we should basically have the same logic as `VRT_String()`.

Skipped reviewing the rest of this function

Reply:
That case is covered in lines 403-411 further down. The idea in L393 is that this is likely to be a very common special case -- there was only one string in the STRANDS object.

And I would have said that the part in lines 393 to 412 are very similar to the same part of ``VRT_String()``, after changing the varags and ``vrt_magic_string`` stuff for STRANDS.

Maybe more so than it needs to be in fact, because this part could be simplified, and that might involve moving the L393 case further down. I'll try a simplification.